File: Solutions/0503/0503.py
# -*- coding:utf-8 -*-
# Author: washing
# DateTime: 2021/3/6 22:35
# File: 0503.py
# Desc: 
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ret = [-1]*n
        stk = []

        for i in range(n*2-1):
            idx = i % n
            while stk and nums[stk[-1]] < nums[idx]:
                ret[stk.pop()] = nums[idx]
            stk.append(idx)

        return ret

File: Solutions/0503/0503.java
/**
 * @Author: washing
 * @DateTime: 2021/03/6 22:34
 * @File: 0503.java
 * @Desc: 
 **/

class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int len = nums.length;
        int[] ret = new int[len];
        Deque<Integer> stk = new LinkedList<Integer>();
        Arrays.fill(ret, -1);

        for (int i=0; i<2*len-1; i++){
            int idx = i % len;
            while (!stk.isEmpty() && nums[stk.peek()] < nums[idx]){
                ret[stk.pop()] = nums[idx];
            }
            stk.push(idx);
        }
        return ret;
    }
}